---
title: "nanostring_geomx_proteogenomics_protein_sept6"
author: "eric allenspach"
date: "2023-10-10"
output: html_document
---

```{r}
#BiocManager::install(version="3.17")

#BiocManager::install("NanoStringNCTools")
#BiocManager::install("GeomxTools")
#BiocManager::install("GeoMxWorkflows")
```

```{r}
library(NanoStringNCTools)
library(GeomxTools)
library(GeoMxWorkflows)
library(here)
```

#loading the data
```{r}
datadir <- here("20230331_Allenspach_AllSeq")
```

```{r}
# automatically list files in each directory for use
DCCFiles <- dir(file.path(datadir, "dccs"), pattern = ".dcc$",
                full.names = TRUE, recursive = TRUE)
PKCFiles <- dir(file.path(datadir, "pkcs"), pattern = ".pkc$",
                                full.names = TRUE, recursive = TRUE)
SampleAnnotationFile <- dir(file.path(datadir, "annotation"), pattern = ".xlsx$",
        full.names = TRUE, recursive = TRUE)
```

```{r}
# Load the data
ourData <- suppressWarnings(readNanoStringGeoMxSet(dccFiles = DCCFiles,
                           pkcFiles = PKCFiles,
                           phenoDataFile = SampleAnnotationFile,
                           phenoDataSheet = "Template",
                           phenoDataDccColName = "Sample_ID",
                           protocolDataColNames = c("aoi", "roi"),
                           configFile = NULL,
                           analyte = "protein",
                           phenoDataColPrefix = "",
                           experimentDataColNames = NULL))
```

```{r}
proteinData <- setSegmentQCFlags(ourData, qcCutoffs = list(percentSaturation = 45,
                                                               minSegmentReads=1000, 
                                                               percentAligned=80, 
                                                               minNegativeCount=10, 
                                                               maxNTCCount=60, 
                                                               minNuclei=16000, 
                                                               minArea=20))

# low sequenced ROIs
lowSaturation <- which(as.data.frame(protocolData(proteinData)[["QCFlags"]])["LowSaturation"] == TRUE)

# remove low quality ROIs
passedQC <- proteinData[, -lowSaturation]
dim(proteinData)
```

```{r}
igg.names <- iggNames(proteinData)
igg.names
```


```{r}
fig <- qcProteinSignal(object = proteinData, neg.names = igg.names)

proteinOrder <- qcProteinSignalNames(object = proteinData, neg.names = igg.names)
genesOfInterest <- c(which(proteinOrder == "Tyrosine Hydroxylase"),
                     which(proteinOrder == "ApoA-I"),
                     which(proteinOrder == "EpCAM"))

fig()
rect(xleft = 0, xright = 4, 
     ybottom = -2, ytop = 2, density = 0, col = "#1B9E77", lwd = 2)
rect(xleft = genesOfInterest[1]-1, xright = genesOfInterest[1]+1, 
     ybottom = -2, ytop = 1.25, density = 0, col = "#D95F02", lwd = 2)
rect(xleft = genesOfInterest[2]-1, xright = genesOfInterest[2]+1, 
     ybottom = -1, ytop = 3, density = 0, col = "#66A61E", lwd = 2)
rect(xleft = genesOfInterest[3]-1, xright = genesOfInterest[3]+1, 
     ybottom = -3, ytop = 6.5, density = 0, col = "#E7298A", lwd = 2)
```
```{r}
proteinOrder <- qcProteinSignalNames(object = proteinData, neg.names = igg.names)

P62 <- which(proteinOrder == "P62")

fig()
```



```{r}
plotConcordance(object = proteinData, targetList = igg.names, plotFactor = "group")
```

```{r}
normfactors <- computeNormalizationFactors(object = proteinData,
                                           area = "AOI.Size.um2",
                                           nuclei = "Nuclei.Counts")

plotNormFactorConcordance(object = proteinData, plotFactor = "group",
                          normfactors = normfactors)
```

```{r}
#HK normalization
#proteinData <- normalize(proteinData, norm_method="hk", toElt = "hk_norm")

#Background normalization
proteinData <- normalize(proteinData, norm_method="neg", toElt = "neg_norm")

#Quantile normalization
proteinData <- normalize(proteinData, norm_method="quant", desiredQuantile = .75, toElt = "q_norm")

names(proteinData@assayData)
```

```{r}
# visualize segments
boxplot(exprs(proteinData),
        col = "#9EDAE5", main = "Raw Counts",
        log = "y", names = 1:95, xlab = "Segment",
        ylab = "Counts, Raw")
```

dimension reductionality
```{r}
library(umap)
library(Rtsne)

# update defaults for umap to contain a stable random_state (seed)
custom_umap <- umap::umap.defaults
custom_umap$random_state <- 42
# run UMAP
umap_out <-
    umap(t(log2(assayDataElement(proteinData , elt = "q_norm"))),  
         config = custom_umap)
pData(proteinData)[, c("UMAP1", "UMAP2")] <- umap_out$layout[, c(1,2)]
ggplot(pData(proteinData),
       aes(x = UMAP1, y = UMAP2, color = region, shape = class)) +
    geom_point(size = 3) +
    theme_bw()
```

```{r}
library(umap)
library(Rtsne)

# update defaults for umap to contain a stable random_state (seed)
custom_umap <- umap::umap.defaults
custom_umap$random_state <- 42
# run UMAP
umap_out <-
    umap(t(log2(assayDataElement(proteinData , elt = "q_norm"))),  
         config = custom_umap)
pData(proteinData)[, c("UMAP1", "UMAP2")] <- umap_out$layout[, c(1,2)]
ggplot(pData(proteinData),
       aes(x = UMAP1, y = UMAP2, color = region, shape = class)) +
    geom_point(size = 3) +
    theme_bw()
```


```{r}
# convert test variables to factors
pData(proteinData)$testClass <-
    factor(pData(proteinData)$class, c("normal", "disease"))
pData(proteinData)$testRegion <- 
    factor(pData(proteinData)$region, c("A", "B"))
pData(proteinData)[["slide"]] <- 
    factor(pData(proteinData)[["slide name"]])
assayDataElement(object = proteinData, elt = "log_q") <-
    assayDataApply(proteinData, 2, FUN = log, base = 2, elt = "q_norm")
    
# run LMM:
# formula follows conventions defined by the lme4 package
results2 <- c()
for(region in c("Control Bone Marrow B", "SETP6 GOF")) {
    ind <- pData(proteinData)$region == region
    mixedOutmc <-
        mixedModelDE(proteinData[, ind],
                     elt = "log_q",
                     modelFormula = ~ testClass + (1 | slide),
                     groupVar = "testClass",
                     nCores = parallel::detectCores(),
                     multiCore = FALSE)
    r_test <- do.call(rbind, mixedOutmc["lsmeans", ])
    tests <- rownames(r_test)
    r_test <- as.data.frame(r_test)
    r_test$Contrast <- tests
    r_test$Gene <- 
        unlist(lapply(colnames(mixedOutmc),
                      rep, nrow(mixedOutmc["lsmeans", ][[1]])))
    r_test$Subset <- region
    r_test$FDR <- p.adjust(r_test$`Pr(>|t|)`, method = "fdr")
    r_test <- r_test[, c("Gene", "Subset", "Contrast", "Estimate", 
                         "Pr(>|t|)", "FDR")]
    results2 <- rbind(results2, r_test)
}
```

```{r}
knitr::kable(subset(results2, proteinOrder %in% goi & Subset == "tubule"), digits = 3,
      caption = "DE results for Genes of Interest",
      align = "lc", row.names = FALSE)
```

Separate healthy and disease for clustering
```{r}
dim(proteinData)

fData(proteinData)

annotation(proteinData)

svarLabels(proteinData)

head(sData(proteinData))

assayData(proteinData)

norm_proteinData <- subset(proteinData, select = phenoData(proteinData)[["class"]] == "normal")
dis_proteinData <- subset(proteinData, select = phenoData(proteinData)[["class"]] == "disease")
```

Run Umaps
```{r}
# update defaults for umap to contain a stable random_state (seed)
custom_umap <- umap::umap.defaults
custom_umap$random_state <- 42
# run UMAP
umap_out <-
    umap(t(log2(assayDataElement(norm_proteinData , elt = "q_norm"))),  
         config = custom_umap)
pData(norm_proteinData)[, c("UMAP1", "UMAP2")] <- umap_out$layout[, c(1,2)]
ggplot(pData(norm_proteinData),
       aes(x = UMAP1, y = UMAP2, color = region, shape = class)) +
    geom_point(size = 3) +
    theme_bw()

# run UMAP
umap_out <-
    umap(t(log2(assayDataElement(dis_proteinData , elt = "q_norm"))),  
         config = custom_umap)
pData(dis_proteinData)[, c("UMAP1", "UMAP2")] <- umap_out$layout[, c(1,2)]
ggplot(pData(dis_proteinData),
       aes(x = UMAP1, y = UMAP2, color = region, shape = class)) +
    geom_point(size = 3) +
    theme_bw()
```

What makes these different in the disease?

Add expression data to umaps (quantitively color each protein marker onto the umap)
```{r}
library(esetVis)

print(esetSpectralMap(eset = norm_proteinData))

ggplot(pData(norm_proteinData),
       aes(x = UMAP1, y = UMAP2, color = assayData(norm_proteinData), shape = class)) +
    geom_point(size = 3) +
    theme_bw()
```
They have an option to do this with tSNE, but not umap. That's too bad, I'll have to do it by hand.

```{r}
library(tidyverse)
library(janitor)

pData(dis_proteinData)[["UMAP1"]]
pData(dis_proteinData)[["UMAP2"]]
sampleNames(dis_proteinData)

umap_dis <- tibble(sample_name = sampleNames(dis_proteinData),
       umap1 = pData(dis_proteinData)[["UMAP1"]],
       umap2 = pData(dis_proteinData)[["UMAP2"]])


fData(dis_proteinData)[["TargetName"]]

dis_sample_read_counts <- assayDataElement(dis_proteinData, elt = "q_norm") %>%
  as_tibble(rownames = "surface_marker") %>%
  pivot_longer(cols = starts_with("DSP"), names_to = "sample_name", values_to = "reads")

dis_df <- left_join(umap_dis, dis_sample_read_counts, by = "sample_name") %>%
  pivot_wider(names_from = surface_marker, values_from = reads) %>%
  clean_names()
```

```{r}
dis_df %>%
  ggplot(aes(x = umap1, y = umap2, color = cd3)) +
  geom_point() +
  scale_color_viridis_c()
```

```{r}
left_join(umap_dis, dis_sample_read_counts, by = "sample_name") %>%
  clean_names() %>%
  ggplot(aes(x = umap1, y = umap2, color = reads)) +
  geom_point() +
  scale_color_viridis_c() +
  facet_wrap(~surface_marker)

left_join(umap_dis, dis_sample_read_counts, by = "sample_name") %>%
  clean_names() %>%
  group_split(surface_marker) %>%
  map(~ggplot(., aes(x = umap1, y = umap2, color = reads)) +
        geom_point() +
        scale_color_viridis_c() +
        labs(title = str_c("surface marker: ", .$surface_marker[1]))
      )
```
```{r}
umap_norm <- tibble(sample_name = sampleNames(norm_proteinData),
       umap1 = pData(norm_proteinData)[["UMAP1"]],
       umap2 = pData(norm_proteinData)[["UMAP2"]])


fData(norm_proteinData)[["TargetName"]]

norm_sample_read_counts <- assayDataElement(norm_proteinData, elt = "q_norm") %>%
  as_tibble(rownames = "surface_marker") %>%
  pivot_longer(cols = starts_with("DSP"), names_to = "sample_name", values_to = "reads")

norm_df <- left_join(umap_norm, norm_sample_read_counts, by = "sample_name") %>%
  pivot_wider(names_from = surface_marker, values_from = reads) %>%
  clean_names()

left_join(umap_norm, norm_sample_read_counts, by = "sample_name") %>%
  clean_names() %>%
  group_split(surface_marker) %>%
  map(~ggplot(., aes(x = umap1, y = umap2, color = reads)) +
        geom_point() +
        scale_color_viridis_c() +
        labs(title = str_c("surface marker: ", .$surface_marker[1]))
      )
```


Want to assign clusters back to transcriptome data

k-means clustering
"Quick approach:
If you don't have cluster assignment already, you can quickly run kmeans clustering to partition cells into two populations.
kmeans(umap_coordinates, k = 2)
```{r}
# disease
kmeans_dis <- umap_dis %>%
  select(umap1, umap2) %>%
  kmeans(centers = 2)

dis_df$cluster <- kmeans_dis$cluster

# norm
kmeans_norm <- umap_norm %>%
  select(umap1, umap2) %>%
  kmeans(centers = 2)

norm_df$cluster <- kmeans_norm$cluster
```


Once you have each region assigned to a cluster, you can run a Wilcox rank sum test for each protein, testing expression differences between the clusters. That will give you a ranked list of proteins with fold-change (FC) and significance. You can consider those proteins with large positive FC as markers of regions in cluster 1, and those with large negative FC as markers of cluster 2. 
wilcox.test(x = cluster1_exp, y = cluster2_exp, paired = F, alternative = "two.sided")

Trying out one before iterating
```{r}
test <- dis_df %>%
  pivot_longer(cols = 4:23, names_to = "surface_marker", values_to = "adj_reads") %>%
  dplyr::filter(surface_marker == "cd4")

toutput <- wilcox.test(x = test$adj_reads[test$cluster == 1], y = test$adj_reads[test$cluster == 2], paired = FALSE, alternative = "two.sided")

toutput$statistic
```


```{r}
dis_wil <- dis_df %>%
  pivot_longer(cols = 4:23, names_to = "surface_marker", values_to = "adj_reads") %>%
  group_split(surface_marker) %>%
  map(function(surf_prot) {
    output <- wilcox.test(x = surf_prot$adj_reads[surf_prot$cluster == 1], y = surf_prot$adj_reads[surf_prot$cluster == 2], paired = FALSE, alternative = "two.sided")
    
    tibble(surface_marker = surf_prot$surface_marker[1],
           statistic = output$statistic,
           parameter = output$parameter,
           p_value = output$p.value,
           null_value = output$null.value,
           alternative = output$alternative,
           method = output$method)
  }) %>%
  bind_rows()

norm_wil <- norm_df %>%
  pivot_longer(cols = 4:23, names_to = "surface_marker", values_to = "adj_reads") %>%
  group_split(surface_marker) %>%
  map(function(surf_prot) {
    output <- wilcox.test(x = surf_prot$adj_reads[surf_prot$cluster == 1], y = surf_prot$adj_reads[surf_prot$cluster == 2], paired = FALSE, alternative = "two.sided")
    
    tibble(surface_marker = surf_prot$surface_marker[1],
           statistic = output$statistic,
           parameter = output$parameter,
           p_value = output$p.value,
           null_value = output$null.value,
           alternative = output$alternative,
           method = output$method)
  }) %>%
  bind_rows()
```

Where cluster1_exp is a vector with the normalized protein expression values for regions in cluster 1 and cluster2_exp is same thing for cluster 2. You'll want to iterate the test for each protein and then assemble results in a dataframe for comparison.
```{r}
dis_wil
norm_wil
```


 Let me know if you want more detailed sample code."